src/clib/solve_rate_cool_g-cpp.C

```diff
@@ -0,0 +1,1012 @@
+        //     `enforce_max_heatcool_subcycle_dt_` & within energy update loop
+        //  5. stop modifying `itmask` in `select_chem_scheme_update_masks_`
+        //  6. remove declaration of `itmask_tmp`, logic that initializes, and
+        //     the loop that uses it to override `itmask`
+
+        // Update dtit (the current subcycle timestep) to ensure it doesn't
+        // exceed the max timestep for cooling/heating
+        // -> zones that will use Newton-Raphson scheme are ignored
+        enforce_max_heatcool_subcycle_dt_(
+          dtit.data(), idx_range, dt, ttot.data(), itmask.data(), tgas.data(),
+          p2d.data(), edot.data(), my_chemistry
+        );
+
+        // Update total and gas energy
+        // -> zones that will use Newton-Raphson scheme are ignored
+        if (my_chemistry->with_radiative_cooling == 1)  {
+          for (int i = idx_range.i_start; i < idx_range.i_stop; i++) {
+            if (itmask[i] != MASK_FALSE) {
+              e(i,j,k) = e(i,j,k) + (gr_float)(edot[i]/d(i,j,k)*dtit[i]);
+            }
+          }
+        }
+
+        if (my_chemistry->primordial_chemistry > 0)  {
+
+          // Solve rate equations with one linearly implicit Gauss-Seidel
+          // sweep of a backward Euler method (for all cells specified by
+          // itmask)
+
+          f_wrap::step_rate_g(
+            dtit.data(), idx_range, anydust, spsolvbuf.h2dust, rhoH.data(),
+            spsolvbuf.dedot_prev, spsolvbuf.HIdot_prev, itmask.data(),
+            itmask_metal.data(), imetal, my_chemistry, my_fields,
+            *my_uvb_rates, spsolvbuf.grain_growth_rates,
+            spsolvbuf.species_tmpdens, spsolvbuf.kcr_buf, spsolvbuf.kshield_buf
+          );
+
+          // Solve rate equations with one linearly implicit Gauss-Seidel
+          // sweep of a backward Euler method (for all cells specified by
+          // itmask_nr)
+
+          f_wrap::step_rate_newton_raphson(
+            imetal, idx_range, iter, dom, chunit, dx_cgs, c_ljeans,
+            dtit.data(), p2d.data(), tgas.data(), tdust.data(),
+            metallicity.data(), dust2gas.data(), rhoH.data(), mmw.data(),
+            spsolvbuf.h2dust, edot.data(), anydust, spsolvbuf.itmask_nr,
+            itmask_metal.data(), spsolvbuf.imp_eng, my_chemistry, my_rates,
+            my_fields, *my_uvb_rates, internalu, grain_temperatures,
+            logTlininterp_buf, cool1dmulti_buf, coolingheating_buf,
+            spsolvbuf.chemheatrates_buf
+          );
+
+        }
+
+        // restore the values of itmask from the backup that we made earlier
+        for (int i = idx_range.i_start; i < idx_range.i_stop; i++) {
+          itmask[i] = itmask_tmp[i];
+        }
+
+        // Add the timestep to the elapsed time for each cell and find
+        //  minimum elapsed time step in this row
+        ttmin = huge8;
+        for (int i = idx_range.i_start; i < idx_range.i_stop; i++) {
+          ttot[i] = std::fmin(ttot[i] + dtit[i], dt);
```

Comment:
Maybe this is brittle or dangerous, but I wonder if all this should be inside an itmask[i] check. If we're here, there should always be at least one true itmask value still and we can save the time of looping over the whole array. This is for the future, but putting this here to make a record of it.